pkg/manager/member/shared_store_failover.go

```diff
@@ -0,0 +1,129 @@
+		if controller.IsIgnoreError(err) {
+			return nil
+		}
+		return err
+	}
+	return nil
+}
+
+func (ssf *sharedStoreFailover) tryMarkAStoreAsFailure(tc *v1alpha1.TidbCluster, failoverPeriod time.Duration) error {
+	ns := tc.GetNamespace()
+	tcName := tc.GetName()
+
+	for storeID, store := range ssf.storeAccess.GetStores(tc) {
+		podName := store.PodName
+		if store.LastTransitionTime.IsZero() {
+			continue
+		}
+		if !ssf.isPodDesired(tc, podName) {
+			// we should ignore the store record of deleted pod, otherwise the
+			// record of deleted pod may be added back to failure stores
+			// (before it enters into Offline/Tombstone state)
+			continue
+		}
+		deadline := store.LastTransitionTime.Add(failoverPeriod)
+		exist := false
+		for _, failureStore := range ssf.storeAccess.GetFailureStores(tc) {
+			if failureStore.PodName == podName {
+				exist = true
+				break
+			}
+		}
+		if store.State == v1alpha1.TiKVStateDown && time.Now().After(deadline) {
+			maxFailoverCount := ssf.storeAccess.GetMaxFailoverCount(tc)
+			if maxFailoverCount != nil && *maxFailoverCount > 0 {
+				ssf.storeAccess.SetFailoverUIDIfAbsent(tc)
+				if !exist {
+					ssf.storeAccess.CreateFailureStoresIfAbsent(tc)
+					if len(ssf.storeAccess.GetFailureStores(tc)) >= int(*maxFailoverCount) {
+						klog.Warningf("%s/%s %s failure stores count reached the limit: %d", ns, tcName, ssf.storeAccess.GetMemberType(), maxFailoverCount)
+						return nil
+					}
+					ssf.storeAccess.SetFailureStores(tc, storeID, v1alpha1.TiKVFailureStore{
+						PodName:   podName,
+						StoreID:   store.ID,
+						CreatedAt: metav1.Now(),
+					})
+					msg := fmt.Sprintf("store[%s] is Down", store.ID)
+					ssf.deps.Recorder.Event(tc, corev1.EventTypeWarning, unHealthEventReason, fmt.Sprintf(unHealthEventMsgPattern, ssf.storeAccess.GetMemberType(), podName, msg))
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (ssf *sharedStoreFailover) isPodDesired(tc *v1alpha1.TidbCluster, podName string) bool {
+	ordinals := ssf.storeAccess.GetStsDesiredOrdinals(tc, true)
+	ordinal, err := util.GetOrdinalFromPodName(podName)
+	if err != nil {
+		klog.Errorf("unexpected pod name %q: %v", podName, err)
+		return false
+	}
+	return ordinals.Has(ordinal)
+}
+
+func (ssf *sharedStoreFailover) RemoveUndesiredFailures(tc *v1alpha1.TidbCluster) {
+	for key, failureStore := range ssf.storeAccess.GetFailureStores(tc) {
+		if !ssf.isPodDesired(tc, failureStore.PodName) {
+			// If we delete the pods, e.g. by using advanced statefulset delete
+			// slots feature. We should remove the record of undesired pods,
+			// otherwise an extra replacement pod will be created.
+			delete(ssf.storeAccess.GetFailureStores(tc), key)
+		}
+	}
+}
+
+func (ssf *sharedStoreFailover) Recover(tc *v1alpha1.TidbCluster) {
+	ssf.storeAccess.ClearFailStatus(tc)
+	klog.Infof("%s recover: clear FailureStores, %s/%s", ssf.storeAccess.GetMemberType(), tc.GetNamespace(), tc.GetName())
+}
```

Comment:
Can we add
```suggestion
}

var _ Failover = (*sharedStoreFailover)(nil)
```
to enforce compiler check that `sharedStoreFailover` is an implementation of `Failover`? Might also be useful if we can have a `NewFailover` which takes `StoreCommonAccess` as input and returns a `Failover` instance (similar to `NewTiKVFailover`).